modules/data.atmosphere/R/tdm_predict_subdaily_function.R

```diff
@@ -0,0 +1,256 @@
+                ], nrow = length(rows.beta), ncol = ncol(betas_nc))
+            ncdf4::nc_close(betas_nc)
+            dat.pred <- predict.met(newdata = dat.temp, model.predict = mod.save, 
+                Rbeta = Rbeta, resid.err = FALSE, model.resid = NULL, Rbeta.resid = NULL, 
+                n.ens = n.ens)
+            
+            #----- Now we do a little quality control per variable
+            
+            # Make Sure that Shortwave is above 0
+            if (v == "surface_downwelling_shortwave_flux_in_air") {
+                dat.pred[dat.pred < 0] <- 0
+            }
+            
+            # Precipitation Re-distribute negative probabilities -- add randomly to
+            # make more peaky If there's no rain on this day, skip the
+            # re-proportioning
+            if (v == "precipitation_flux") {
+                if (max(dat.pred) > 0) {
+                  tmp <- 1:nrow(dat.pred)  # A dummy vector of the 
+                  for (j in 1:ncol(dat.pred)) {
+                    if (min(dat.pred[, j]) >= 0) 
+                      next
+                    rows.neg <- which(dat.pred[, j] < 0)
+                    rows.add <- sample(tmp[!tmp %in% rows.neg], length(rows.neg), 
+                      replace = TRUE)
+                    
+                    for (z in 1:length(rows.neg)) {
+                      dat.pred[rows.add[z], j] <- dat.pred[rows.add[z], 
+                        j] - dat.pred[rows.neg[z], j]
+                      dat.pred[rows.neg[z], j] <- 0
+                    }
+                  }
+                  dat.pred <- dat.pred/rowSums(dat.pred)
+                  dat.pred[is.na(dat.pred)] <- 0
+                }
+                # Convert precip into real units
+                dat.pred <- dat.pred * as.vector((dat.temp$precipitation_flux.day))
+            }
+            
+            # Longwave needs some sanity bounds
+            if (v == "surface_downwelling_longwave_flux_in_air") {
+                dat.pred <- dat.pred^2  # because squared to prevent negative numbers
+                dat.pred[dat.pred < 100] <- 100
+                dat.pred[dat.pred > 600] <- 600
+            }
+            
+            # Specific Humidity sometimes ends up with high or infinite values
+            if (v == "specific_humidity") {
+                if (max(dat.pred) > 0.03) {
+                  specific_humidity.fix <- ifelse(quantile(dat.pred, 0.99) < 
+                    0.03, quantile(dat.pred, 0.99), 0.03)
+                  dat.pred[dat.pred > specific_humidity.fix] <- specific_humidity.fix
+                }
+            }
+            # ---------- End Quality Control
+            
+            # ---------- Begin propogating values and saving values
+            # Shortwave Radiaiton
+            if (v == "surface_downwelling_shortwave_flux_in_air") {
+                # Randomly pick which values to save & propogate
+                cols.prop <- sample(1:n.ens, ncol(dat.sim[[v]]), replace = TRUE)
```

Comment:
@mdietze this is a really good point, I need to think about this for a bit. Since we only do 1 variable at a time, I may need generate a list of which values to propagate beforehand and read that from the list so it's consistent across variables